src/shared/sleep-config.c

```diff
@@ -102,6 +102,94 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+int battery_is_low(void) {
+        _cleanup_free_ char *bat_cap_level = NULL;
+        int r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity_level", &bat_cap_level);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity_level: %m");
+
+        log_debug("Current battery level: %s", bat_cap_level);
+
+        return STR_IN_SET(bat_cap_level, "Critical", "Low");
+}
+
+int read_battery_capacity_percentage(void) {
+        _cleanup_free_ char *bat_cap = NULL;
+        int battery_capacity, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to parse battery capacity: %m");
+
+        if (battery_capacity < 0 || battery_capacity > 100)
+               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
+
+        log_debug("Current battery charge percentage: %d%%", battery_capacity);
+
+        return battery_capacity;
+}
+
+static int get_battery_discharge_rate_filepath(char **ret_path) {
+        sd_id128_t machine_id;
+        int r;
+
+        assert(ret_path);
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0) {
+                return log_debug_errno(r, "Failed to get machine ID: %m");
+        }
+
+        *ret_path = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_percentage_rate_per_hour");
+        if (!ret_path) {
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to parse battery discharge percentage rate file path: %m");
```

Comment:
this should just be return log_oom_debug() as it's not really parsing anything, if this fails it's because of an out of memory situation (and again no brackets)